File: contest/biweekly-contest-61/5859.0_Count_Number_of_Pairs_With_Absolute_Difference_K.py
'''
AC
Brute Force
T: O(N^2)
S: O(1)
'''
class Solution:
    def countKDifference(self, nums: List[int], k: int) -> int:
        ans, N = 0, len(nums)
        for i in range(N - 1):
            for j in range(i + 1, N):
                if abs(nums[i] - nums[j]) == k:
                    ans += 1
        return ans

'''
Hash Table
T: O(N)
S: O(N)

执行用时：40 ms, 在所有 Python3 提交中击败了91.35% 的用户
内存消耗：15 MB, 在所有 Python3 提交中击败了45.83% 的用户
通过测试用例：236 / 236
'''
class Solution:
    def countKDifference(self, nums: List[int], k: int) -> int:
        ans, c = 0, Counter(nums)
        for x, cnt in c.items():
            ans += cnt * c[x + k]
        return ans

File: cpp/5859.0_Count_Number_of_Pairs_With_Absolute_Difference_K.cpp
/*
Hash Table

执行用时：8 ms, 在所有 C++ 提交中击败了97.44% 的用户
内存消耗：12.1 MB, 在所有 C++ 提交中击败了21.86% 的用户
通过测试用例：236 / 236
*/

class Solution {
public:
    int countKDifference(vector<int>& nums, int k) {
        vector<int> cnt(100);
        for (int n: nums) {
            cnt[n - 1]++;
        }
        int ans = 0;
        for (int i = 0; i + k < 100; ++i) {
            ans += cnt[i] * cnt[i + k];
        }
        return ans;
    }
};




/*
Hash Table
使用int cnt[100] = {0};会更省内存

执行用时：8 ms, 在所有 C++ 提交中击败了97.44% 的用户
内存消耗：12 MB, 在所有 C++ 提交中击败了73.63% 的用户
通过测试用例：236 / 236
*/
class Solution {
public:
    int countKDifference(vector<int>& nums, int k) {
        // vector<int> cnt(100);
        int cnt[100] = {0};
        for (int n: nums) {
            cnt[n - 1]++;
        }
        int ans = 0;
        for (int i = 0; i + k < 100; ++i) {
            ans += cnt[i] * cnt[i + k];
        }
        return ans;
    }
};


/*
Brute Force

执行用时：16 ms, 在所有 C++ 提交中击败了70.57% 的用户
内存消耗：12 MB, 在所有 C++ 提交中击败了64.71% 的用户
通过测试用例：236 / 236
*/
class Solution {
public:
    int countKDifference(vector<int>& nums, int k) {
        int ans = 0, len = nums.size();
        for (int i = 0; i < len; ++i) {
            for (int j = i + 1; j < len; ++j) {
                if (fabs(nums[i] - nums[j]) == k) ans++;
            }
        }
        return ans;
    }
};
